### PR TITLE
60287 pass subscriber configuration from Reader to Handler

### PR DESCRIPTION
**Pull request by dawidkeshop**

### CODE REVIEW

File: src/CaptainHook.DirectorService/DirectorService.cs [hunk 1]
```diff
@@ -81,29 +82,7 @@ protected override async Task RunAsync(CancellationToken cancellationToken)
...
-                                PartitionSchemeDescription = new SingletonPartitionSchemeDescription(),
-                                ServiceTypeName = ServiceNaming.EventReaderServiceType,
-                                ServiceName = new Uri(readerServiceNameUri),
-                                InitializationData = initializationData,
-                                PlacementConstraints = _defaultServiceSettings.DefaultPlacementConstraints
-                            },
-                            TimeSpan.FromSeconds(30),
-                            cancellationToken);
-                    }
+                    await CreateServiceForSubscriber(subscriber, serviceList, cancellationToken);
```
Comment (tabman83) [inline src/CaptainHook.DirectorService/DirectorService.cs:85]: This has been discussed on a separate project - as per MS conventions we should consider adding the -Async suffix to any async method to improve clarity.

Comment (dawidkeshop) [reply src/CaptainHook.DirectorService/DirectorService.cs:85]: Sure

File: src/CaptainHook.Common/MessageData.cs [hunk 2]
```diff
@@ -51,5 +53,14 @@ public MessageData(string payload, string type, string subscriberName, string re
...
 
+        /// <summary>
+        /// The subscriber configuration related to the message.
+        /// </summary>
+        public SubscriberConfiguration SubscriberConfig { get; set; }
+
+        /// <summary>
+        /// The Webhook configuration related to the message.
+        /// </summary>
+        public WebhookConfig WebhookConfig { get; set; }
```
Comment (mszymeshop) [inline src/CaptainHook.Common/MessageData.cs:64]: We can probably get rid of that and start relying on `SubscriberConfig` entirely. To be added as a Tech Debt as discussed offline.

File: src/CaptainHook.Common/ServiceModels/EventReaderInitData.cs [hunk 3]
```diff
@@ -41,6 +45,19 @@ public static EventReaderInitData FromSubscriberConfiguration(SubscriberConfigur
             };
         }
 
+        public static EventReaderInitData FromSubscriberConfiguration(SubscriberConfiguration subscriberConfiguration, WebhookConfig webhookConfig)
```
Comment (mszymeshop) [inline src/CaptainHook.Common/ServiceModels/EventReaderInitData.cs:48]: The above method is probably no longer needed or both can be merged together.

File: src/CaptainHook.EventHandlerActor/Handlers/EventHandlerFactory.cs [hunk 4]
```diff
@@ -57,10 +51,10 @@ public IHandler CreateEventHandler(string eventType, string subscriberName)
                     _authenticationHandlerFactory,
                     _requestLogger,
                     _bigBrother,
-                    subscriberConfig);
+                    messageData.SubscriberConfig);
             }
 
-            return CreateWebhookHandler(subscriberConfig.Name);
+            return CreateWebhookHandler(messageData.WebhookConfig);
```
Comment (mszymeshop) [inline src/CaptainHook.EventHandlerActor/Handlers/EventHandlerFactory.cs:57]: Potentially you could just pass the WebhookConfig from the subscriber here. If that is out of scope let's include it in the tech debt which was discussed.

File: src/CaptainHook.Common/MessageData.cs (hunk 2 above)
Comment (chomick) [reply src/CaptainHook.Common/MessageData.cs:64]: Yes, we will have to refactor the models over time but would minimise changes for now.

File: src/CaptainHook.DirectorService/DirectorService.cs [hunk 5]
```diff
@@ -144,6 +124,67 @@ protected override async Task RunAsync(CancellationToken cancellationToken)
...
+                   TimeSpan.FromSeconds(30),
+                   cancellationToken);
+            }
+        }
+
+        private static (string newName, string oldName) FindServiceNames(ICollection<string> serviceList, string readerServiceNameUri)
+        {
+            string newName = $"{readerServiceNameUri}-a", oldName = null;
+
+            if (serviceList.Contains(readerServiceNameUri))
```
Comment (mszymeshop) [inline src/CaptainHook.DirectorService/DirectorService.cs:163]: I would consider simplifying the algorithm. My suggestion: find the first occurrence of `uri, uri-b, uri-a` in `serviceList` and use it as `oldName`. If `oldName` matches `newName` use `uri-b` instead as newName. What do you think @dawidkeshop ?

Comment (dawidkeshop) [reply src/CaptainHook.DirectorService/DirectorService.cs]: Yes, you're right. Especially that there is possibility that more than one service should be deleted.

**Comment by mszymeshop** [comment]
After the compatibility commit, it looks good ✔️
